types/cli-table/index.d.ts

```diff
@@ -1,45 +1,55 @@
 // Type definitions for cli-table 0.3
 // Project: https://github.com/Automattic/cli-table
 // Definitions by: AryloYeung <https://github.com/arylo>
+//                 Antoni Spaanderman <https://github.com/antonilol>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.1
+// TypeScript Version: 4.9
 
-interface TableOptions {
-    chars: Partial<Record<(
-        'top' |
-        'top-mid' |
-        'top-left' |
-        'top-right' |
-        'bottom' |
-        'bottom-mid' |
-        'bottom-left' |
-        'bottom-right' |
-        'left' |
-        'left-mid' |
-        'mid' |
-        'mid-mid' |
-        'right' |
-        'right-mid' |
-        'middle'
-    ), string>>;
-    truncate: string;
-    colors: boolean;
-    colWidths: number[];
-    colAligns: Array<"left" | "middle" | "right">;
-    style: Partial<{
-        'padding-left': number;
-        'padding-right': number;
-        head: string[];
-        border: string[];
-        compact: boolean;
-    }>;
-    head: string[];
+interface TableOptions<T> {
+    rows?: T[];
+    chars?: {
+        [k in
+            | 'top'
+            | 'top-mid'
+            | 'top-left'
+            | 'top-right'
+            | 'bottom'
+            | 'bottom-mid'
+            | 'bottom-left'
+            | 'bottom-right'
+            | 'left'
+            | 'left-mid'
+            | 'mid'
+            | 'mid-mid'
+            | 'right'
+            | 'right-mid'
+            | 'middle']?: string;
+    };
+    truncate?: string;
+    colors?: boolean;
+    colWidths?: number[];
+    colAligns?: Array<'left' | 'middle' | 'right'>;
+    style?: {
+        'padding-left'?: number;
+        'padding-right'?: number;
+        head?: string[];
+        border?: string[];
+        compact?: boolean;
+    };
+    head?: string[];
 }
 
-declare class Table extends Array {
-    constructor(options?: Partial<TableOptions>);
+type HorizontalTableRow = string[];
+type VerticalTableRow = Record<string, string>;
+type CrossTableRow = Record<string, string[]>;
+type TableRow = HorizontalTableRow | VerticalTableRow | CrossTableRow;
+
+declare class Table<T extends TableRow> extends Array<T> {
```

Comment:
I think you need to have a default for the new `T` type parameters (same for the new `T` above in `TableOptions<T>`), so that existing code that used the old types don't break, e.g. `var x: TableOptions;`. Ideally it should be the same default, since likely if someone declared an `option: TableOptions`, they want to use it to create an object of class `Table`. Since `T` has a constraint (`TableRow`), I'm not sure what the default for `T` should be.

Reply:
having the type parameter the constraint will work on places we try to infer type arguments, e.g. `new Table()`, but it won't in places where we're using just the types and the type arguments must be passed in unless they have a default, such as:
`declare const opts: TableOptions;`, so I think you still need to update the type parameters on `Table` and `TableOptions` to have a default value, e.g. `declare class Table<T extends TableRow=TableRow>`.

Reply:
makes sense, fixed